pkg/endpoint/endpoint.go

```diff
@@ -1181,27 +1182,29 @@ func (e *Endpoint) Update(owner Owner, opts models.ConfigurationMap) error {
 
 	// Option changes may be overridden by the policy configuration.
 	// Currently we return all-OK even in that case.
-	changed, ctCleaned, err := e.TriggerPolicyUpdatesLocked(owner, opts)
+	_, ctCleaned, err := e.TriggerPolicyUpdatesLocked(owner, opts)
 	if err != nil {
 		e.Mutex.Unlock()
 		ctCleaned.Wait()
 		return UpdateCompilationError{err.Error()}
 	}
 
-	if changed {
-		changed = e.SetStateLocked(StateWaitingToRegenerate, "Updated endpoint options; policy changes apply to this endpoint")
-	}
-	e.Mutex.Unlock()
-	ctCleaned.Wait()
-
```

Comment:
It should not be necessary to regenerate if the endpoint options did not actually change. Maybe the fix should be that:

1. set `changed = true` if `opts == nil` so that the we try to regenerate if called from the API to force regeneration.
2. SetStateLocked is only called `changed == true` like before.
3. `e.Regenerate()` is only called if `changed == true` and `SetStateLocked()` succeeded.
- It is no error if endpoint options did not really change
- It is an error if SetStateLocked() failed.

Reply:
@jrajahalme We want to regenerate the endpoint even if the options did not change - that was the reason for the change https://github.com/cilium/cilium/pull/3040/commits/a98ddad1a447e331190979a65d8cc3fda81963ef .

Reply:
My proposal would keep that behavior, note the setting of `changed` to `true` when `opts ==nil` (the step 1). This forces *trying* to regenerate in that case too, but only if the state change succeeds. Also, of the state change fails the operation should fail. Basically a98ddad is wrong in the sense that it should only have added the (attempted) regeneration if `opts == nil`. Now it tries to regenerate also in cases where options were given, but nothing actually changed.

Reply:
I'm fine to keep the behaviour of not regenerating if nothing changed and set `changed = true` on `opts == nil`

Reply:
> note the setting of changed to true when opts ==nil (the step 1). This forces trying to regenerate in that case too, but only if the state change succeeds.

With this patch though, we don't care if opts are nil or if `changed` is true or false; the opts are passed into `TriggerPolicyUpdates` either way. The call to regenerate on line 1203 is always made without regard to whether `opts ==  nil` or not.

> Now it tries to regenerate also in cases where options were given, but nothing actually changed.
We do want to regenerate even if nothing actually changed.

Reply:
Ahhh, never mind, I understand now. In the case that opts are provided, do not regenerate if not necessary, but if they are not provided, then always regenerate. Got it!